Add caching, tracing, logging, and batching to the plugin

craigtmoore opened the pull request:
Adds caching using Caffine cache, logging with java.util.logging, tracing using opentelemtry, and inserts the data into the database using batching. We have some builds on our Jenkins instance that have over 29,000 test results so we added these changes to optimize performance. 

Fixes #412

- Update DatabaseTestResultStorage:
  * add caseResultsCache
  * add packageResultsCache
  * update publish() method to use batch updates to improve performance
  * Add @WithSpan annotation to each method
  * add logger
- Update DatabaseTestResultStorageTest:
  * add basic unit-test with a mocked database
  * move duplicate code segments to constants and methods
  * print tables with proper column widths (and truncate values that are too large)
- Update pom.xml:
  * add opentelemtry dependencies (for tracing)
  * add caffine dependency (for caching)
  * add dependency on `kotlin-std-lib-jkd8` to fix dependency resolution issues
  * tie the `hpi` goal to the `compile` phase
- update docker-compose.yaml:
  * add commented out 'mysql' database config
  * add jaeger server
  * add jenkins server with otel agent embedded
- docker-compose-with-zipkin.yaml: same as docker-compose.yaml, but using zipkin instead of jaeger
- Add Dockerfile: creates an instance of the Jenkins docker image that includes the opentelemetry agent jar file, used to generate theweatherman/jenkins:lts-jdk17-otel image.
- Add setup_jenkins.sh: bash script for installing the plugin and configuring jenkins to store junit results in the databse (works with the docker-compose.yaml)

### Testing done

I re-used the existing tests to verify that I've introduced no new regressions, but also addeed a new unit-test to the `getCaseResults_mockDatabase()` method and verify that it loads the data correctly.

### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [NA] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

timja commented on README.md:
````diff
@@ -22,14 +22,79 @@ Tables will be automatically created.
+Here's an example of how you use the setup script:
+
+```
+./setup_jenkins.sh http://<hostname>:8080 <api_token>
````
why not do this in the Dockerfile, 

Something like:
```
# install the plugin from update center to get all dependencies
RUN jenkins-plugin-cli --plugins junit-sql-storage configuration-as-code
# override with locally built version
COPY target/(whatever it is).hpi /usr/share/jenkins/ref/plugins/junit-sql-storage.jpi 

# configuration as code config for configuring this plugin
COPY docker/jenkins.yaml /usr/share/jenkins/ref/jenkins.yaml
```

craigtmoore replied:
Wow thats great, I didn't know about the `jenkins-plugin-cli` binary or I would have done it this way :-/ I tried it out and it works fine.

timja commented on README.md:
````diff
@@ -60,6 +129,52 @@ unclassified:
+You can also query the mysql (or postgres) database by connecting to the `db` container (note: `winpty` is only required when running on Windows git bash)
+
+```
+$ winpty docker exec -it junit-sql-storage-plugin-db-1 mysql --password=root mysql
````
winpty can be mentioned separately, shouldn't be in the default example:
```suggestion
$ docker compose exec db mysql --password=root mysql
```

craigtmoore replied:
Good point, I've removed it from the example.

timja commented on docker-compose-with-zipkin.yaml:
```diff
@@ -0,0 +1,29 @@
+
+services:
+  db:
+    image: mysql:8.0.37-debian
```
I don't think it needs to be this specific:

what about either:
```suggestion
    image: mysql:lts
```

or:
```suggestion
    image: mysql:8.0-debian
```

craigtmoore replied:
I think I'll just remove that file. It doesn't really add anything.

timja commented on docker-compose-with-zipkin.yaml:
```diff
@@ -0,0 +1,29 @@
+      - 9411:9411
+
+  jenkins:
+    image: theweatherman/jenkins:otel
```
Can be use the official Jenkins image please especially since we're already copying in the agent?

craigtmoore replied:
I've removed this file so I guess this topic is moot.

timja commented on docker-compose-with-zipkin.yaml:
```diff
@@ -0,0 +1,29 @@
+
+  jenkins:
+    image: theweatherman/jenkins:otel
+    ports:
```
Can be use the official Jenkins image please especially since we're already copying in the agent?

Shouldn't this have a build statement:

```suggestion
    build:
      context: .
    ports:
```

timja commented on pom.xml:
```diff
@@ -109,22 +114,60 @@
+            <artifactId>opentelemetry-instrumentation-annotations</artifactId>
+            <version>2.2.0</version>
+        </dependency>
+        <dependency>
+            <groupId>io.opentelemetry</groupId>
+            <artifactId>opentelemetry-api</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.opentelemetry</groupId>
+            <artifactId>opentelemetry-sdk-trace</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.opentelemetry</groupId>
+            <artifactId>opentelemetry-sdk-metrics</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.opentelemetry</groupId>
+            <artifactId>opentelemetry-exporter-otlp</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.opentelemetry</groupId>
+            <artifactId>opentelemetry-exporter-logging</artifactId>
+        </dependency>
```
these imports should all be replaced by a dependency on the opentelemetry plugin so that multiple versions aren't bundled into Jenkins causing problems:
https://plugins.jenkins.io/opentelemetry/dependencies/

craigtmoore replied:
I tried to do as you suggested, but that plugin is not available at
* https://repo.jenkins-ci.org/public/org/jenkins-ci/plugins/

Do you know the artifact details for the maven pom.xml?

timja replied:
Its on the page I linked

craigtmoore replied:
Thanks, I was able to make it work, but I had to replace all the @WithSpan annotations that I had in the Java code and use manual tracing instead 🤷‍♂️

timja commented on pom.xml:
```diff
@@ -109,22 +114,60 @@
+            <artifactId>opentelemetry-exporter-logging</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>com.github.ben-manes.caffeine</groupId>
+            <artifactId>caffeine</artifactId>
+            <version>3.1.8</version>
```
replace with api plugin:
```suggestion
            <groupId>io.jenkins.plugins</groupId>
            <artifactId>caffeine-api</artifactId>
```

craigtmoore replied:
Done

timja commented on pom.xml:
```diff
@@ -109,22 +114,60 @@
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-core</artifactId>
+            <version>5.12.0</version>
```
version is set by bom in the parent pom: https://github.com/jenkinsci/plugin-pom/blob/master/pom.xml#L177
```suggestion
```

craigtmoore replied:
OK, I've removed it.

timja commented on pom.xml:
```diff
@@ -137,4 +180,20 @@
+                        <goals>
+                            <goal>hpi</goal>
+                        </goals>
+                        <phase>compile</phase>
```
why? this doesn't seem right

craigtmoore replied:
When I tried doing `mvn hpi:hpi` for the first time it always failed so I moved it to the compile phase so it would come after the code is compiled.

timja replied:
can you raise an issue at https://github.com/jenkinsci/maven-hpi-plugin/

I don't understand that plugin well enough but don't really want this here, I've never had the need to call `mvn hpi:hpi` directly.

Either:
`mvn clean package` or `mvn hpi:run` should be sufficient and work fine

craigtmoore replied:
OK, I've removed that section from the pom.xml and adjusted the ./deploy.sh file.

timja commented on setup_jenkins.sh:
```diff
@@ -0,0 +1,174 @@
+#! /bin/bash
+
+# Colorized output colors
```
I think this file can be completely replaced by just putting the configuration into the Dockerfile as explained in the README comment

craigtmoore replied:
I agree, I've removed it.

timja reviewed (commented):
How extensively have you tested this? Have you deployed it to your instance with the large number of results?

My main concern is that the optimised database calls for only reading e.g the fail or skipped count are now being in Java so the full amount needs to be read into memory and filtered client side.

Its possible that most of the methods get called when this is used anyway so it may not impact much, or possibly there could either be more caches.

I assume you found a cache was needed because retrieval was slow? do you have any info on this? 

The batch changes make full sense

craigtmoore commented on Dockerfile:
```diff
@@ -0,0 +1,11 @@
+RUN chmod 644 /opt/opentelemetry-javaagent.jar
+RUN chown jenkins:jenkins /opt/opentelemetry-javaagent.jar
+
+USER jenkins
```
Thanks, I didn't know this was added. I've updated the Dockerfile.

craigtmoore commented on README.md:
````diff
@@ -22,14 +22,79 @@ Tables will be automatically created.
 
-e.g. you could install the [PostgreSQL Database](https://plugins.jenkins.io/database-postgresql/) plugin.
+```
+$ docker logs -f junit-sql-storage-plugin_jenkins_1
````
Thanks, I've updated it to use `logs` from `docker compose`.

timja commented on README.md:
```diff
@@ -22,14 +22,79 @@ Tables will be automatically created.
 ## Getting started
 
-Install your database vendor specific plugin, you can use the Jenkins plugin site to search for it:
+To install the plugin login into Jenkins → Manage Jenkins → Manage Plugins → Available → Search for 'JUnit SQL Storage' → Install.
```
```suggestion
To install the plugin login to Jenkins → Manage Jenkins → Manage Plugins → Available → Search for 'JUnit SQL Storage' → Install.
```

craigtmoore replied:
OK, I've re-worded it.

timja commented on src/test/java/io/jenkins/plugins/junit/storage/database/DatabaseTestResultStorageTest.java:
```diff
@@ -60,202 +78,269 @@
+
+    @Before
+    public void setup() {
+        Logger logger = Logger.getLogger(DatabaseTestResultStorage.class.getName());
+        logger.setLevel(Level.INFO);
+
+        ConsoleHandler consoleHandler = new ConsoleHandler();
+        consoleHandler.setLevel(Level.INFO);
+        logger.addHandler(consoleHandler);
+
+        // Remove default handlers to prevent duplicate messages
+        Logger rootLogger = Logger.getLogger("");
+        Handler[] handlers = rootLogger.getHandlers();
+        for (Handler handler : handlers) {
+            rootLogger.removeHandler(handler);
+        }
+        rootLogger.addHandler(consoleHandler);
```
Not sure exactly what you were doing here, seems you're setting logging to INFO?

This would probably simplify what you are trying to do: https://github.com/jenkinsci/jenkins-test-harness/blob/master/src/main/java/org/jvnet/hudson/test/LoggerRule.java

craigtmoore replied:
Yep, that's great. Didn't know about it. I've implemented it.

timja commented on src/test/java/io/jenkins/plugins/junit/storage/database/DatabaseTestResultStorageTest.java:
```diff
@@ -60,202 +78,269 @@
+        }
+    }
+
+    public void verifyTableStructure(Connection connection)
```
why is this needed? (and why public?)

craigtmoore replied:
I've made the method private (oops).

craigtmoore commented:
As far as testing goes, the unit-tests at our company are quite large, over 29,000 test results in a single build. We're looking into using a plugin like this, but the performance was quite poor when first tried it (which is why I looked at improving the performance). At first I deployed it using mysql instead of postgres because I could not connect the database to Jenkins.

Once I was able to deploy it using and connect it to a dockerized mysql database, I ran the following pipeline:

```
pipeline {
    agent any
    stages {
        stage('Extract test results') {
            steps {
                sh 'unzip -o /tmp/archive.zip -d build'
            }
        }
    }
    post {
        always {
            junit 'build/**/TEST-*.xml'
        }
    }
}
```

Basically, I downloaded the zip file with the large number of test results from our production Jenkins server and then
copied that to the Jenkins instance, using:
```
docker cp archive.zip junit-sql-storage-plugin-jenkins-1:/tmp/archive.zip
```

Then I ran the pipeline and it took a long time to store the test results (nearly 3 minutes to store the results using
a mysql database). So I added open telemetry to the plugin so I could see where the bottlenecks were. I decided to add
batching to the publish method to help reduce the time it takes to store the values. This certainly helped, but the
other problem was loading the test results took a really long time. I thought it might have been that the queries were
too slow, but it turned out that the junit plugin was calling the `getAllPackageResults` over 1500 times
(which can takes 100s of milliseconds) so I added caching which reduced the processing time down to 10s of microseconds for each method call. I decided to take it one step further and cache the value returned by the `retrieveCaseResults()` method
(which I renamed `getCaseResults()`), this also really helped because the case results query takes ~20ms to run and
that takes a long time when there are over 29000 test results. Also, that query was being called by many of the
 meta-data methods, like:
* getFailedTests
* getSkippedTests
* getPassedTests
By caching the caseResults, we avoided running the query repeatedly and reduced the processing time to 10s of
microseconds.

So to answer your comment, yes I did a lot of testing to verify that my changes improved the performance. The telemetry that I added was also very useful in figuring out the bottle necks.

timja commented on jenkins-config.yaml:
```diff
@@ -0,0 +1,11 @@
+        username: postgres
+        validationQuery: "SELECT 1"
+  junitTestResultStorage:
+    storage: "database"
```
```suggestion
    storage: "database"

```

craigtmoore replied:
I've added a line break to the end of the file.

timja reviewed (commented):
Getting close

craigtmoore commented on README.md:
````diff
@@ -60,6 +114,52 @@ unclassified:
+Here's an example of how to use it
+
+```
+java -jar java-cli.jar -s http://localhost:8080 -u admin -p <api_token> -f junit-sql-storage-plugin-config.yml
````
Good point, I've updated it. It was completely wrong :-P

craigtmoore commented on README.md:
```diff
@@ -60,6 +114,52 @@ unclassified:
+
+### Accessing the database
+
+You can also query the mysql (or postgres) database by connecting to the `db` container.
```
OK

craigtmoore commented on deploy.sh:
```diff
@@ -0,0 +1,28 @@
+  exit 1
+fi
+echo
+# Build Docker Image
+echo -e "${LightGreen}docker build --tag jenkins .${NC}"
+docker build --tag jenkins .
+echo
+# Deploy the docker swarm
+echo -e "${LightGreen}docker compose up -d${NC}"
+docker compose up -d
```
I couldn't get that to work properly. It didn't build the Docker image so I've kept it as a 2 step operation.

craigtmoore commented on docker-compose.yaml:
```diff
@@ -8,3 +8,25 @@ services:
+      OTEL_METRICS_EXPORTER: none
+      OTEL_LOGS_EXPORTER: none
+      OTEL_JAVAAGENT_LOGGING: simple
+      JAVA_OPTS: "-javaagent:/opt/opentelemetry-javaagent.jar"
```
Sounds good. I've remove the opentelemetry-javaagent and added the setup wizard skip property

craigtmoore commented:
I think I've resolved all of your comments, please let me know if there is any thing else. I really appreciate all of the feedback. I'm going to do a bit more testing, especially with junit attachments plugin.

timja reviewed (commented):
This looks really good, I've tested it out, fixed up the deploy script to work with build and a couple of other minor things.

I've found one issue though with the open telemetry integration, I would expect the spans to be linked to the build but they aren't:

<img width="1249" alt="image" src="https://github.com/jenkinsci/junit-sql-storage-plugin/assets/21194782/9852f321-7f36-435b-98ad-37ea3b7ca844">

I'm wondering if the `junit` step needs a `StepHandler`?

https://github.com/jenkinsci/opentelemetry-plugin/blob/03559ec4adf9ed6843e7fb0635a5e4cc9f961fb2/src/main/java/io/jenkins/plugins/opentelemetry/job/step/DurableTaskHandler.java

see also:
https://github.com/jenkinsci/opentelemetry-plugin/issues/850

I couldn't see any `TRACEPARENT` env values being set

cc @cyrille-leclerc or @kuisathaverat if you can provide any advice

timja commented on src/main/java/io/jenkins/plugins/junit/storage/database/DatabaseTestResultStorage.java:
```diff
@@ -114,60 +156,96 @@ private static class RemotePublisherImpl implements RemotePublisher {
-                                statement.setNull(12, Types.VARCHAR);
+        @Override
+        public void publish(TestResult result, TaskListener listener) throws IOException {
+            var publishSpan = createSpan("DatabaseTestResultStorage.RemotePublisherImpl.publish");
```
As far as I can tell this only works for agents on the controller not remote agents

timja replied:
Setting `-Dotel.java.global-autoconfigure.enabled=true` on the agent seems to work

timja commented:
I don't think its `StepHandler` I think if the variables are propagated through createRemotePublisher and this code is copied then it would work:
https://github.com/jenkinsci/opentelemetry-plugin/blob/82e3e1b2574a68864dcd05b43b95d0aeb7f41249/src/main/java/io/jenkins/plugins/opentelemetry/job/OtelEnvironmentContributorService.java#L41-L52

timja commented:
Continued at https://github.com/jenkinsci/junit-sql-storage-plugin/pull/414

jonesbusy commented:
Shoudn't the opentelemetry plugin be optional ? What about jenkins instance using SQL storage but not connected to an Opentelemetry collector ?

From what I remember few month ago is that having OpenTelemetry plugin installed will display some deak links on job/build page when the plugin is not configured

timja commented:
Possibly, would be better to remove those dead links I think if the plugin isn't configured at all.

Plugins should be able to integrate without it adding a bunch of things to the UI

cyrille-leclerc commented:
Thanks for your interest in the opentelemetry plugin.
I think we could make it optional for simpler integration of otel in the other Jenkins plugins, I have a few checks to do. 
I'll get back to you ASAP

timja commented:
just raised a PR https://github.com/jenkinsci/opentelemetry-plugin/pull/868

kuisathaverat commented:
> I couldn't see any TRACEPARENT env values being set

you have to enable the export of the environment variables that define the TRACEPARENT, but I not sure how this plugin get the Opentelemetry context, it probably has to retrieve it from the job not front and environment variable. We store the context in actions in the job.

timja commented:
I was able to get it in https://github.com/jenkinsci/junit-sql-storage-plugin/pull/414

cyrille-leclerc commented:
@craigtmoore @timja the way you integrate OpenTelemetry is great. `GlobalOpenTelemetry.get()` is the way to go.
I'm looking at some improvements to this pattern `GlobalOpenTelemetry.get()`. I'll get back to you shortly.

I saw @timja' PR, Ill review it it asap, I have to understand the challenge you faced :-)
* https://github.com/jenkinsci/opentelemetry-plugin/pull/868

timja commented:
Anything running on the controller I think would be quite straightforward, but running on agents is a bit more complicated, its not the nicest / cleanest implementation in https://github.com/jenkinsci/junit-sql-storage-plugin/pull/414
